numba/core/pythonapi.py

```diff
@@ -1150,6 +1150,23 @@ def string_from_kind_and_data(self, kind, string, size):
         fn = self._get_function(fnty, name=fname)
         return self.builder.call(fn, [kind, string, size])
 
+    def bytes_as_string(self, obj):
+        fnty = ir.FunctionType(self.cstring, [self.pyobj])
+        fname = "PyBytes_AsString"
+        fn = self._get_function(fnty, name=fname)
+        return self.builder.call(fn, [obj])
+
+    def bytes_as_string_and_size(self, obj, p_buffer, p_length):
+        fnty = ir.FunctionType(
+            self.pyobj,
+            [self.pyobj, self.cstring.as_pointer(), self.py_ssize_t.as_pointer()],
+        )
+        fname = "PyBytes_AsStringAndSize"
+        fn = self._get_function(fnty, name=fname)
+        result = self.builder.call(fn, [obj, p_buffer, p_length])
+        ok = self.builder.icmp_unsigned("!=", Constant(result.type, None), result)
```

Comment:
In case of error, [`PyBytes_AsStringAndSize`](https://docs.python.org/3/c-api/bytes.html#c.PyBytes_AsStringAndSize) returns `-1`.

```suggestion
        ok = self.builder.icmp_unsigned("!=", Constant(result.type, -1), result)
```

Reply:
Sorry, I should have tested it before committing it - `-1` causes it to fail at runtime:
```python
>               raise RuntimeError("LLVM IR parsing error\n{0}".format(errmsg))
E               RuntimeError: Failed in nopython mode pipeline (step: native lowering)
E               LLVM IR parsing error
E               <string>:69:24: error: integer constant must have integer type
E                 %".42" = icmp ne i8* -1, %".41"
E                                      ^

```

Reply:
My bad as well! The return type must be changed from `pyobj` to `py_ssize_t`. Also, I think the comparison instruction should be a signed integer comparison (icmp_signed)

Reply:
Hi @ianna, from the [docs](https://docs.python.org/3/c-api/bytes.html#c.PyBytes_AsStringAndSize), `PyBytes_AsStringAndSize` returns an integer. My suggestion is to change the return type in the signature to match CPython:

```python
        fnty = ir.FunctionType(
            ir.IntType(32),
            [self.pyobj, self.cstring.as_pointer(), self.py_ssize_t.as_pointer()],
        )
```